glib/src/gstring.rs

```diff
@@ -14,6 +14,326 @@ use std::ptr;
+    pub unsafe fn from_ptr<'a>(ptr: *const c_char) -> &'a Self {
+        let cstr = CStr::from_ptr(ptr);
+        Self::from_bytes_with_nul_unchecked(cstr.to_bytes_with_nul())
+    }
+    // rustdoc-stripper-ignore-next
+    /// Converts this GLib string to a byte slice containing the trailing 0 byte.
+    ///
+    /// This function is the equivalent of [`GStr::to_bytes`] except that it will retain the
+    /// trailing nul terminator instead of chopping it off.
+    #[inline]
+    pub fn to_bytes_with_nul(&self) -> &[u8] {
+        &self.0
+    }
+    // rustdoc-stripper-ignore-next
+    /// Converts this GLib string to a byte slice.
+    ///
+    /// The returned slice will **not** contain the trailing nul terminator that this GLib
+    /// string has.
+    #[inline]
+    pub fn to_bytes(&self) -> &[u8] {
+        &self.0[0..self.0.len() - 1]
+    }
+    // rustdoc-stripper-ignore-next
+    /// Returns the inner pointer to this GLib string.
+    ///
+    /// The returned pointer will be valid for as long as `self` is, and points to a contiguous
+    /// region of memory terminated with a 0 byte to represent the end of the string.
+    ///
+    /// **WARNING**
+    ///
+    /// The returned pointer is read-only; writing to it (including passing it to C code that
+    /// writes to it) causes undefined behavior. It is your responsibility to make
+    /// sure that the underlying memory is not freed too early.
+    #[inline]
+    pub fn as_ptr(&self) -> *const c_char {
+        self.0.as_ptr() as *const _
+    }
+    // rustdoc-stripper-ignore-next
+    /// Converts this GLib string to a string slice.
+    #[inline]
+    pub fn as_str(&self) -> &str {
+        unsafe { std::str::from_utf8_unchecked(self.to_bytes()) }
+    }
+    // rustdoc-stripper-ignore-next
+    /// Converts this GLib string to a C string slice.
+    #[inline]
+    pub fn as_c_str(&self) -> &CStr {
+        unsafe { CStr::from_bytes_with_nul_unchecked(&self.0) }
+    }
+}
+
+// rustdoc-stripper-ignore-next
+/// Converts a static string literal into a static nul-terminated string.
+///
+/// The expanded expression has type [`&'static GStr`]. This macro will panic if the
+/// string literal contains any interior nul bytes.
+///
+/// # Examples
+///
+/// ```
+/// # fn main() {
+/// use glib::{gstr, GStr, GString};
+///
+/// const MY_STRING: &GStr = gstr!("Hello");
+/// assert_eq!(MY_STRING.to_bytes_with_nul()[5], 0u8);
+/// let owned: GString = MY_STRING.to_owned();
+/// assert_eq!(MY_STRING, owned);
+/// # }
+/// ```
+///
+/// [`&'static GStr`]: crate::GStr
+#[macro_export]
+macro_rules! gstr {
+    ($s:literal) => {
+        unsafe { $crate::GStr::from_bytes_with_nul_unchecked($crate::cstr_bytes!($s)) }
+    };
+}
+
+impl Default for &GStr {
+    fn default() -> Self {
+        const SLICE: &[c_char] = &[0];
+        unsafe { GStr::from_ptr(SLICE.as_ptr()) }
+    }
+}
+
+impl<'a> From<&'a CStr> for &'a GStr {
+    fn from(s: &'a CStr) -> Self {
+        s.to_str().unwrap();
```

Comment:
You could directly transmute the return value here btw, but also this should probably rather be a `TryFrom`?

Reply:
Yes, a lot of these for `GString` should also be `TryFrom`. Should I change those too?

Reply:
Let's do that in a follow up, it seems a bit disruptive. For now I just changed this one

Reply:
I made a PR draft, can make an issue too if you want: https://github.com/gtk-rs/gtk-rs-core/pull/600

Reply:
Draft PR is even better :)